File: firecracker_vm/vm.py
"""
Single Firecracker microVM management.

Each FirecrackerVM instance corresponds to exactly one Firecracker OS process.
Communication happens over a Unix domain socket using Firecracker's REST API.

Lifecycle:
    1. Create VMConfig with kernel, rootfs, etc.
    2. Instantiate FirecrackerVM (spawns the firecracker process)
    3. Configure boot source, drives, machine config via API
    4. Start the VM (InstanceStart action)
    5. Communicate with guest over vsock/serial
    6. Snapshot / restore / destroy as needed

For snapshot-restored VMs, the lifecycle starts at step 5 — the VM resumes
exactly where the snapshot was taken, with all guest processes intact.
"""

from __future__ import annotations

import json
import os
import shutil
import socket
import subprocess
import tempfile
import time
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any, Optional
import http.client


class VMState(Enum):
    """Tracks the VM lifecycle from the host's perspective."""
    CREATED = "created"       # Firecracker process started, not yet configured
    CONFIGURED = "configured"  # Boot source + drives set, not yet running
    RUNNING = "running"       # Guest is executing
    PAUSED = "paused"         # Guest frozen (required before snapshot)
    STOPPED = "stopped"       # Process terminated


@dataclass
class VMConfig:
    """
    Everything needed to boot a fresh Firecracker VM.

    Fields:
        firecracker_bin: Path to the firecracker binary
        kernel_path:     Path to the guest kernel (vmlinux)
        rootfs_path:     Path to the root filesystem image (ext4)
        vcpu_count:      Number of virtual CPUs
        mem_size_mib:    Guest memory in MiB
        boot_args:       Kernel command line arguments
        track_dirty:     Enable dirty page tracking for diff snapshots
    """
    firecracker_bin: str = "firecracker"
    kernel_path: str = ""
    rootfs_path: str = ""
    vcpu_count: int = 1
    mem_size_mib: int = 256
    boot_args: str = "console=ttyS0 reboot=k panic=1 pci=off random.trust_cpu=on init=/opt/init.sh"
    track_dirty: bool = True
    extra_drives: list[dict] = field(default_factory=list)


class FirecrackerVM:
    """
    Manages a single Firecracker microVM process.

    This is the lowest-level abstraction. It owns:
    - The firecracker OS process
    - A working directory for socket, logs, etc.
    - The API connection over the Unix socket

    Thread safety: Not thread-safe. Each VM should be managed from a single
    thread. Use VMPool for concurrent VM management.
    """

    def __init__(self, config: VMConfig, work_dir: Optional[str] = None,
                 vm_id: Optional[str] = None):
        self.config = config
        self.vm_id = vm_id or f"vm-{os.getpid()}-{id(self):x}"
        self.state = VMState.CREATED

        # Each VM gets its own working directory to avoid socket/file collisions
        if work_dir:
            self.work_dir = Path(work_dir).resolve()
            self.work_dir.mkdir(parents=True, exist_ok=True)
            self._owns_work_dir = False
        else:
            self.work_dir = Path(tempfile.mkdtemp(
                prefix=f"fc-{self.vm_id}-")).resolve()
            self._owns_work_dir = True

        self.socket_path = self.work_dir / "firecracker.sock"
        self._process: Optional[subprocess.Popen] = None
        self._log_path = self.work_dir / "firecracker.log"
        # Track which rootfs this VM is using (may differ from config for forks)
        self.rootfs_path = config.rootfs_path

    def boot(self) -> None:
        """
        Full boot sequence: spawn process, configure, start.

        Communication with the guest uses vsock (virtio socket) —
        much faster than serial, supports full-bandwidth data transfer.
        """
        self._spawn_process()
        self._configure_machine()
        self._configure_boot_source()
        self._configure_rootfs()
        self._configure_vsock()
        for drive in self.config.extra_drives:
            self._add_drive(drive)
        self._start_instance()

    def _spawn_process(self) -> None:
        """
        Start the Firecracker binary as a child process.

        The process listens on a Unix domain socket for API commands.
        We pipe stdin/stdout for serial console communication with the guest.
        Firecracker connects the guest's /dev/ttyS0 to the process's
        stdin/stdout, giving us a bidirectional byte stream to the guest.
        """
        # Clean up any stale socket
        if self.socket_path.exists():
            self.socket_path.unlink()

        # Resolve paths to absolute since we set cwd to work_dir
        fc_bin = str(Path(self.config.firecracker_bin).resolve())

        self._process = subprocess.Popen(
            [
                fc_bin,
                "--api-sock", str(self.socket_path),
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=open(self._log_path, "w"),
            cwd=str(self.work_dir),  # vsock "v.sock" resolves relative to this
        )

        # Wait for the socket to become available
        for _ in range(100):  # 10 second timeout
            if self.socket_path.exists():
                try:
                    self._api_call("GET", "/")
                    return
                except Exception:
                    pass
            time.sleep(0.1)

        raise RuntimeError(
            f"Firecracker failed to start. Check {self._log_path}"
        )

    def _spawn_process_for_restore(self) -> None:
        """
        Start a Firecracker process that will be used for snapshot restore.

        Same as _spawn_process but the VM won't be configured from scratch —
        instead, the caller will load a snapshot into it.
        """
        self._spawn_process()

    def _api_call(self, method: str, path: str,
                  body: Optional[dict] = None) -> dict[str, Any]:
        """
        Make an HTTP request to the Firecracker API over the Unix socket.

        Firecracker's API is a simple REST interface:
        - PUT for creating/setting resources (idempotent)
        - PATCH for modifying resources
        - GET for reading state

        Returns the parsed JSON response body, or raises on error.

        Under the hood this uses Python's http.client with a Unix socket
        connection. In a Rust implementation, you'd use hyper with a
        UnixStream transport — same idea, different plumbing.
        """
        conn = _UnixHTTPConnection(str(self.socket_path))
        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        body_str = json.dumps(body) if body else None

        conn.request(method, path, body=body_str, headers=headers)
        response = conn.getresponse()
        response_body = response.read().decode("utf-8")
        conn.close()

        if response.status >= 400:
            raise RuntimeError(
                f"Firecracker API error {response.status} on {method} {path}: "
                f"{response_body}"
            )

        if response_body:
            try:
                return json.loads(response_body)
            except json.JSONDecodeError:
                return {}
        return {}

    def _configure_machine(self) -> None:
        """Set vCPU count, memory, and dirty page tracking."""
        self._api_call("PUT", "/machine-config", {
            "vcpu_count": self.config.vcpu_count,
            "mem_size_mib": self.config.mem_size_mib,
            "track_dirty_pages": self.config.track_dirty,
        })

    def _configure_boot_source(self) -> None:
        """Set the guest kernel and boot arguments."""
        kernel = str(Path(self.config.kernel_path).resolve())
        self._api_call("PUT", "/boot-source", {
            "kernel_image_path": kernel,
            "boot_args": self.config.boot_args,
        })

    def _configure_rootfs(self) -> None:
        """Attach the root filesystem as a virtio-blk device."""
        rootfs = str(Path(self.config.rootfs_path).resolve())
        self._api_call("PUT", "/drives/rootfs", {
            "drive_id": "rootfs",
            "path_on_host": rootfs,
            "is_root_device": True,
            "is_read_only": False,
        })

    def _configure_vsock(self) -> None:
        """
        Attach a virtio-vsock device for host↔guest communication.

        We use a RELATIVE uds_path ("v.sock") so that each Firecracker
        process resolves it relative to its own cwd (the VM's work_dir).
        This is critical for snapshot/restore: the snapshot bakes the
        uds_path, so if we used an absolute path, all restored children
        would try to bind to the same socket and collide.
        """
        self._api_call("PUT", "/vsock", {
            "vsock_id": "1",
            "guest_cid": 3,
            "uds_path": "v.sock",
        })

    @property
    def vsock_uds_path(self) -> str:
        """Absolute path to the vsock UDS (resolved from work_dir + relative path)."""
        return str(self.work_dir / "v.sock")

    @property
    def serial_stdin(self):
        """
        Writable pipe to the guest's /dev/ttyS0 (via Firecracker's stdin).

        Whatever you write here appears on the guest's serial input.
        """
        if self._process is None:
            raise RuntimeError("VM process not started")
        return self._process.stdin

    @property
    def serial_stdout(self):
        """
        Readable pipe from the guest's /dev/ttyS0 (via Firecracker's stdout).

        Whatever the guest writes to /dev/ttyS0 appears here. Note: this
        includes kernel boot messages, init output, AND agent protocol data.
        The client must parse through the noise to find protocol lines.
        """
        if self._process is None:
            raise RuntimeError("VM process not started")
        return self._process.stdout

    def _add_drive(self, drive_config: dict) -> None:
        """Attach an additional virtio-blk device."""
        drive_id = drive_config.get("drive_id", "extra")
        self._api_call("PUT", f"/drives/{drive_id}", drive_config)

    def _start_instance(self) -> None:
        """Send InstanceStart action — this boots the guest kernel."""
        self._api_call("PUT", "/actions", {
            "action_type": "InstanceStart",
        })
        self.state = VMState.RUNNING

    def pause(self) -> None:
        """
        Pause the VM — freezes all vCPUs.

        This is a prerequisite for snapshot creation. The guest doesn't
        know it's paused; from its perspective, time simply stops.
        """
        if self.state != VMState.RUNNING:
            raise RuntimeError(
                f"Can only pause a running VM, got {self.state}")
        self._api_call("PATCH", "/vm", {"state": "Paused"})
        self.state = VMState.PAUSED

    def resume(self) -> None:
        """
        Resume a paused VM — unfreezes all vCPUs.

        From the guest's perspective, time jumps forward by however long
        the VM was paused.
        """
        if self.state != VMState.PAUSED:
            raise RuntimeError(
                f"Can only resume a paused VM, got {self.state}")
        self._api_call("PATCH", "/vm", {"state": "Resumed"})
        self.state = VMState.RUNNING

    def create_snapshot(self, snapshot_dir: str,
                        snapshot_type: str = "Full") -> "Snapshot":
        """
        Create a snapshot of the current VM state.

        Produces two files:
        - vmstate: serialized CPU registers, device state, KVM state
        - memory:  full guest RAM contents (or dirty pages for Diff)

        The VM must be paused before calling this. After snapshot creation,
        the VM can be resumed and continues running normally.

        The disk (rootfs) is NOT included in the snapshot — the caller
        must manage disk file copies separately if needed for forking.

        Returns a Snapshot object that can be used to restore new VMs.
        """
        if self.state != VMState.PAUSED:
            raise RuntimeError("VM must be paused before snapshotting")

        from firecracker_vm.snapshot import Snapshot

        snap_dir = Path(snapshot_dir).resolve()
        snap_dir.mkdir(parents=True, exist_ok=True)
        vmstate_path = snap_dir / "vmstate"
        memory_path = snap_dir / "memory"

        self._api_call("PUT", "/snapshot/create", {
            "snapshot_type": snapshot_type,
            "snapshot_path": str(vmstate_path),
            "mem_file_path": str(memory_path),
        })

        return Snapshot(
            vmstate_path=str(vmstate_path),
            memory_path=str(memory_path),
            rootfs_path=str(Path(self.config.rootfs_path).resolve()),
            config=self.config,
        )

    def destroy(self) -> None:
        """
        Kill the Firecracker process and clean up.

        This is the nuclear option — the guest gets no chance to shut down
        gracefully. For our use case (disposable RL environment VMs), this
        is fine.
        """
        if self._process and self._process.poll() is None:
            self._process.kill()
            self._process.wait(timeout=5)
        self.state = VMState.STOPPED

        # Clean up socket
        if self.socket_path.exists():
            try:
                self.socket_path.unlink()
            except OSError:
                pass

    def is_alive(self) -> bool:
        """Check if the Firecracker process is still running."""
        if self._process is None:
            return False
        return self._process.poll() is None

    @classmethod
    def restore_from_snapshot(
        cls,
        snapshot: "Snapshot",
        work_dir: Optional[str] = None,
        vm_id: Optional[str] = None,
        rootfs_path: Optional[str] = None,
        resume: bool = True,
    ) -> "FirecrackerVM":
        """
        Create a new VM by restoring from a snapshot.

        This is the core of the forking mechanism. The new VM:
        1. Gets a fresh Firecracker process
        2. Loads the snapshot's vmstate and memory (MAP_PRIVATE / CoW)
        3. Optionally resumes immediately

        The rootfs_path parameter allows using a copy-on-write copy of
        the original disk, which is important for forking — each child
        needs its own writable disk.

        Performance note: The memory file is mmap'd MAP_PRIVATE, so only
        pages the child modifies are physically copied. This makes
        fork+restore very fast (~30ms) and memory-efficient.
        """
        from firecracker_vm.snapshot import Snapshot

        config = snapshot.config
        vm = cls(config, work_dir=work_dir, vm_id=vm_id)
        vm._spawn_process_for_restore()

        # Use provided rootfs or the snapshot's original
        actual_rootfs = rootfs_path or snapshot.rootfs_path
        vm.rootfs_path = actual_rootfs

        vm._api_call("PUT", "/snapshot/load", {
            "snapshot_path": snapshot.vmstate_path,
            "mem_backend": {
                "backend_type": "File",
                "backend_path": snapshot.memory_path,
            },
            "enable_diff_snapshots": config.track_dirty,
            "resume_vm": resume,
        })

        vm.state = VMState.RUNNING if resume else VMState.PAUSED
        return vm

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.destroy()

    def __repr__(self):
        return f"FirecrackerVM(id={self.vm_id}, state={self.state.value})"


class _UnixHTTPConnection(http.client.HTTPConnection):
    """
    HTTP connection over a Unix domain socket.

    Firecracker's API server listens on a Unix socket, not TCP.
    This adapter lets us use Python's standard http.client with it.

    The trick: we override connect() to create a Unix socket and
    connect to the file path, but everything else (HTTP framing,
    request/response parsing) works identically to TCP.
    """

    def __init__(self, socket_path: str, timeout: float = 10.0):
        # The host parameter doesn't matter for Unix sockets,
        # but http.client requires it
        super().__init__("localhost", timeout=timeout)
        self._socket_path = socket_path

    def connect(self):
        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.sock.settimeout(self.timeout)
        self.sock.connect(self._socket_path)

File: src/vm.rs
use anyhow::{bail, Context, Result};
use http_body_util::{BodyExt, Full};
use hyper::body::Bytes;
use hyper::Request;
use hyper_util::rt::TokioIo;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::time::Duration;
use tokio::net::UnixStream;
use tokio::process::{Child, Command};

use crate::snapshot::Snapshot;

/// VM lifecycle state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VMState {
    Created,
    Running,
    Paused,
    Stopped,
}

/// Configuration for booting a fresh Firecracker VM.
#[derive(Debug, Clone)]
pub struct VMConfig {
    pub firecracker_bin: PathBuf,
    pub kernel_path: PathBuf,
    pub rootfs_path: PathBuf,
    pub vcpu_count: u32,
    pub mem_size_mib: u32,
    pub boot_args: String,
    pub track_dirty: bool,
}

impl Default for VMConfig {
    fn default() -> Self {
        Self {
            firecracker_bin: PathBuf::from("firecracker"),
            kernel_path: PathBuf::new(),
            rootfs_path: PathBuf::new(),
            vcpu_count: 1,
            mem_size_mib: 128,
            boot_args: [
                "console=ttyS0",
                "reboot=k",
                "panic=1",
                "pci=off",
                "random.trust_cpu=on", // avoid entropy starvation
                "init=/opt/init.sh",
                "mitigations=off", // skip spectre/meltdown — not needed for benchmarking
                "audit=0",         // disable audit subsystem
                "quiet",           // reduce kernel log noise
                "loglevel=1",      // only critical kernel messages
                "nomodule",        // skip module loading
            ]
            .join(" "),
            track_dirty: true,
        }
    }
}

/// Manages a single Firecracker microVM process.
///
/// Each instance owns one Firecracker OS process and communicates
/// with it over a Unix domain socket (HTTP API via hyper).
pub struct FirecrackerVM {
    pub config: VMConfig,
    pub state: VMState,
    pub work_dir: PathBuf,
    pub rootfs_path: PathBuf,
    socket_path: PathBuf,
    log_path: PathBuf,
    process: Option<Child>,
}

impl FirecrackerVM {
    pub fn new(config: VMConfig, work_dir: Option<PathBuf>, vm_id: &str) -> Result<Self> {
        let work_dir = match work_dir {
            Some(d) => {
                std::fs::create_dir_all(&d)?;
                std::fs::canonicalize(&d).context("canonicalize work_dir")?
            }
            None => {
                let td = tempfile::Builder::new()
                    .prefix(&format!("fc-{}-", vm_id))
                    .tempdir()?;
                let d = td.path().to_path_buf();
                std::mem::forget(td);
                d
            }
        };
        std::fs::create_dir_all(&work_dir)?;
        let rootfs_path = std::fs::canonicalize(&config.rootfs_path)
            .unwrap_or_else(|_| config.rootfs_path.clone());

        Ok(Self {
            config,
            state: VMState::Created,
            socket_path: work_dir.join("firecracker.sock"),
            log_path: work_dir.join("firecracker.log"),
            rootfs_path,
            work_dir,
            process: None,
        })
    }

    /// Full boot: spawn process, configure, start.
    pub async fn boot(&mut self) -> Result<()> {
        self.spawn_process().await?;
        self.configure_machine().await?;
        self.configure_boot_source().await?;
        self.configure_rootfs().await?;
        self.configure_vsock().await?;
        self.start_instance().await?;
        Ok(())
    }

    /// Spawn the Firecracker process and wait for the API socket.
    ///
    /// After this returns, the process is idle and ready to accept
    /// API calls (boot config or snapshot load). Public so that
    /// `ProcessPool` can pre-warm processes.
    pub async fn spawn_process(&mut self) -> Result<()> {
        let _ = std::fs::remove_file(&self.socket_path);

        let fc_bin = std::fs::canonicalize(&self.config.firecracker_bin)
            .context("resolve firecracker binary path")?;

        let log_file = std::fs::File::create(&self.log_path)?;
        let child = Command::new(&fc_bin)
            .arg("--api-sock")
            .arg(&self.socket_path)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(log_file)
            .current_dir(&self.work_dir)
            .spawn()
            .context("spawn firecracker")?;

        self.process = Some(child);

        // Wait for API socket to become available
        let deadline = tokio::time::Instant::now() + Duration::from_secs(10);
        loop {
            if self.socket_path.exists() && self.call_firecracker("GET", "/", None).await.is_ok() {
                return Ok(());
            }
            if tokio::time::Instant::now() > deadline {
                bail!("Firecracker failed to start. Check {:?}", self.log_path);
            }
            tokio::time::sleep(Duration::from_millis(50)).await;
        }
    }

    /// Make an HTTP request to the Firecracker API over the Unix socket.
    ///
    /// Opens a fresh connection per call. Firecracker's API is
    /// single-threaded and low-frequency — connection reuse isn't
    /// worth the complexity of keeping an idle connection alive.
    async fn call_firecracker(
        &self,
        method: &str,
        path: &str,
        body: Option<Value>,
    ) -> Result<Value> {
        let stream = UnixStream::connect(&self.socket_path)
            .await
            .context("connect to API socket")?;
        let io = TokioIo::new(stream);

        let (mut sender, conn) = hyper::client::conn::http1::handshake(io)
            .await
            .context("HTTP handshake")?;

        // Drive the connection in the background.
        // The spawned task exits when the response is fully read
        // and `sender` is dropped.
        tokio::spawn(async move {
            if let Err(e) = conn.await {
                eprintln!("[api] connection error: {e}");
            }
        });

        let body_bytes = match &body {
            Some(v) => v.to_string().into_bytes(),
            None => Vec::new(),
        };

        let mut builder = Request::builder()
            .method(method)
            .uri(path)
            .header("Host", "localhost")
            .header("Accept", "application/json");

        if body.is_some() {
            builder = builder
                .header("Content-Type", "application/json")
                .header("Content-Length", body_bytes.len());
        }

        let req = builder
            .body(Full::new(Bytes::from(body_bytes)))
            .context("build request")?;

        let response = sender.send_request(req).await.context("send request")?;
        let status = response.status();
        let resp_bytes = response
            .into_body()
            .collect()
            .await
            .context("read response body")?
            .to_bytes();
        let body_text = String::from_utf8_lossy(&resp_bytes);

        if status.as_u16() >= 400 {
            bail!(
                "Firecracker API error {} on {} {}: {}",
                status.as_u16(),
                method,
                path,
                body_text.trim()
            );
        }

        if body_text.trim().is_empty() {
            Ok(Value::Null)
        } else {
            Ok(serde_json::from_str(body_text.trim()).unwrap_or(Value::Null))
        }
    }

    async fn configure_machine(&self) -> Result<()> {
        self.call_firecracker(
            "PUT",
            "/machine-config",
            Some(json!({
                "vcpu_count": self.config.vcpu_count,
                "mem_size_mib": self.config.mem_size_mib,
                "track_dirty_pages": self.config.track_dirty,
            })),
        )
        .await?;
        Ok(())
    }

    async fn configure_boot_source(&self) -> Result<()> {
        let kernel = std::fs::canonicalize(&self.config.kernel_path)?;
        self.call_firecracker(
            "PUT",
            "/boot-source",
            Some(json!({
                "kernel_image_path": kernel.to_str().unwrap(),
                "boot_args": self.config.boot_args,
            })),
        )
        .await?;
        Ok(())
    }

    async fn configure_rootfs(&self) -> Result<()> {
        self.call_firecracker(
            "PUT",
            "/drives/rootfs",
            Some(json!({
                "drive_id": "rootfs",
                "path_on_host": self.rootfs_path.to_str().unwrap(),
                "is_root_device": true,
                "is_read_only": true,
            })),
        )
        .await?;
        Ok(())
    }

    async fn configure_vsock(&self) -> Result<()> {
        self.call_firecracker(
            "PUT",
            "/vsock",
            Some(json!({
                "vsock_id": "1",
                "guest_cid": 3,
                "uds_path": "v.sock",
            })),
        )
        .await?;
        Ok(())
    }

    async fn start_instance(&mut self) -> Result<()> {
        self.call_firecracker(
            "PUT",
            "/actions",
            Some(json!({
                "action_type": "InstanceStart",
            })),
        )
        .await?;
        self.state = VMState::Running;
        Ok(())
    }

    pub async fn pause(&mut self) -> Result<()> {
        if self.state != VMState::Running {
            bail!("Can only pause a running VM, got {:?}", self.state);
        }
        self.call_firecracker("PATCH", "/vm", Some(json!({"state": "Paused"})))
            .await?;
        self.state = VMState::Paused;
        Ok(())
    }

    /// Create a snapshot of this (paused) VM.
    ///
    /// When `memory_path` is `None`, creates a full snapshot: Firecracker
    /// dumps all guest memory to `<snapshot_dir>/memory`. When `memory_path`
    /// is `Some(base)`, creates a diff snapshot: FC writes only pages dirtied
    /// since the last snapshot/restore into `base`, overwriting those offsets
    /// in-place. The caller must ensure `base` already contains a valid full
    /// image so that clean pages retain their correct content.
    ///
    /// After the write, we kick off async writeback via
    /// `sync_file_range(SYNC_FILE_RANGE_WRITE)` to prevent dirty page
    /// accumulation from causing stalls on later iterations.
    pub async fn create_snapshot(
        &self,
        snapshot_dir: &Path,
        memory_path: Option<&Path>,
    ) -> Result<Snapshot> {
        if self.state != VMState::Paused {
            bail!("VM must be paused before snapshotting");
        }
        std::fs::create_dir_all(snapshot_dir)?;
        let snap_dir = std::fs::canonicalize(snapshot_dir)?;
        let vmstate_path = snap_dir.join("vmstate");
        let is_diff = memory_path.is_some();
        let memory_path = memory_path
            .map(|p| p.to_path_buf())
            .unwrap_or_else(|| snap_dir.join("memory"));

        let snapshot_type = if is_diff { "Diff" } else { "Full" };

        self.call_firecracker(
            "PUT",
            "/snapshot/create",
            Some(json!({
                "snapshot_type": snapshot_type,
                "snapshot_path": vmstate_path.to_str().unwrap(),
                "mem_file_path": memory_path.to_str().unwrap(),
            })),
        )
        .await?;

        // Kick off async writeback of the memory file immediately.
        //
        // Firecracker writes guest memory into the page cache. Without
        // this, the kernel accumulates dirty pages and eventually hits
        // dirty_ratio, causing a *future* write/mmap call to block
        // synchronously for writeback — random multi-second stalls.
        //
        // SYNC_FILE_RANGE_WRITE is non-blocking: starts writeback in
        // the background and returns immediately.
        #[cfg(target_os = "linux")]
        {
            use std::os::unix::io::AsRawFd;
            if let Ok(f) = std::fs::File::open(&memory_path) {
                unsafe {
                    libc::sync_file_range(f.as_raw_fd(), 0, 0, libc::SYNC_FILE_RANGE_WRITE);
                }
            }
        }

        Ok(Snapshot {
            vmstate_path,
            memory_path,
            rootfs_path: self.rootfs_path.clone(),
            config: self.config.clone(),
        })
    }

    /// Load a snapshot into an already-spawned Firecracker process.
    ///
    /// The process must be freshly spawned (no prior boot or load).
    /// This is the fast path used by the process pool — the spawn
    /// and API socket wait have already happened, so this is just
    /// one HTTP round-trip to Firecracker.
    pub async fn load_snapshot(&mut self, snapshot: &Snapshot, resume: bool) -> Result<()> {
        self.rootfs_path = snapshot.rootfs_path.clone();
        self.call_firecracker(
            "PUT",
            "/snapshot/load",
            Some(json!({
                "snapshot_path": snapshot.vmstate_path.to_str().unwrap(),
                "mem_backend": {
                    "backend_type": "File",
                    "backend_path": snapshot.memory_path.to_str().unwrap(),
                },
                "enable_diff_snapshots": snapshot.config.track_dirty,
                "resume_vm": resume,
            })),
        )
        .await?;

        self.state = if resume {
            VMState::Running
        } else {
            VMState::Paused
        };
        Ok(())
    }

    /// Absolute path to the vsock UDS.
    pub fn vsock_uds_path(&self) -> PathBuf {
        self.work_dir.join("v.sock")
    }

    /// Take ownership of the process's stdout handle (serial console).
    pub fn take_stdout(&mut self) -> Option<tokio::process::ChildStdout> {
        self.process.as_mut().and_then(|p| p.stdout.take())
    }

    /// Kill the Firecracker process and clean up the socket file.
    pub fn destroy(&mut self) {
        if let Some(mut child) = self.process.take() {
            // start_kill() sends SIGKILL — non-async, just a syscall.
            let _ = child.start_kill();
            // The Child is dropped here; tokio reaps the zombie via
            // a background SIGCHLD handler.
        }
        self.state = VMState::Stopped;
        let _ = std::fs::remove_file(&self.socket_path);
    }
}

impl Drop for FirecrackerVM {
    fn drop(&mut self) {
        self.destroy();
    }
}
